pocs/filterwheel/libefw.py

```diff
@@ -0,0 +1,281 @@
+        """Main class representing the ZWO EFW library interface.
+
+        On construction loads the shared object/dynamically linked version of the EFW SDK library,
+        which must be already installed (see https://astronomy-imaging-camera.com/software-drivers).
+
+        The name and location of the shared library can be manually specified with the library_path
+        argument, otherwise the ctypes.util.find_library function will be used to try to locate it.
+
+        Args:
+            library_path (str, optional): path to the library e.g. '/usr/local/lib/libEFWFilter.so'
+
+        Returns:
+            `~pocs.filter.libefw.EFWDriver`
+
+        Raises:
+            pocs.utils.error.NotFound: raised if library_path not given & find_library fails to
+                locate the library.
+            OSError: raises if the ctypes.CDLL loader cannot load the library.
+        """
+        super().__init__(name='EFWFilter', library_path=library_path, **kwargs)
+
+    # Methods
+
+    def get_SDK_version(self):
+        """Get the version for the SDK."""
+        # EFW SDK has no way to access SDK version.
+        return "Unknown"
+
+    def get_devices(self):
+        """Get connected device 'UIDs' and corresponding device nodes/handles/IDs.
+
+        EFW SDK has no way to access any unique identifier for connected filterwheels.
+        Instead we construct an ID from combination of filterwheel name, number of
+        positions and integer ID. This will probably not be deterministic, in general,
+        and is only guaranteed to be unique between multiple filterwheels on a single
+        computer.
+        """
+        n_filterwheels = self.get_num()  # Nothing works if you don't call this first.
+        filterwheels = {}
+        for i in range(n_filterwheels):
+            fw_id = self.get_ID(i)
+            self.open(fw_id)
+            info = self.get_property(fw_id)
+            self.close(fw_id)
+            filterwheels[f"{info['name']}_{info['slot_num']}_{fw_id}"] = fw_id
+        return filterwheels
+
+    def get_num(self):
+        """Get the count of connected EFW filterwheels."""
+        count = self._CDLL.EFWGetNum()
+        self.logger.debug(f"Found {count} connected EFW filterwheels.")
+        return count
+
+    def get_product_ids(self):
+        """Get product IDs of supported(?) EFW filterwheels.
+
+        The SDK documentation does not explain that the product IDs returned by this function are,
+        but from experiment and analogy with a similar function in the ASI camera SDK it appears
+        this is a list of the product IDs of the filterwheels that the SDK supports, not the
+        product IDs of the connected filterwheels. There appears to be no way to obtain the
+        product IDs of the connected filterwheel(s).
+        """
+        n_filterwheels = self._CDLL.EFWGetProductIDs(0)
+        if n_filterwheels > 0:
+            product_ids = (ctypes.c_int * n_filterwheels)()
```

Comment:
A function call here?

Reply:
Yes. `ctypes` can allow for some some funny looking syntax, sometimes.

Here I am creating an array data type on the fly by multiplying a scalar data type (`ctypes.c_int`) by an integer (`n_filterwheels`). That's actually the recommended method: https://docs.python.org/3/library/ctypes.html#arrays. I then immediately create an instance of that array data type by calling it.

I could make this look a bit less weird by creating the array data type and assigning it to a variable on one line, then creating the instance and assigning that to another variable on the next line. I only use the data type once though, so it didn't seem worth storing.